### PR TITLE
Upgrade to core 0.94.1

### PR DESCRIPTION
**Pull request by cmelchior**

### CODE REVIEW

**Comment by beeender** [comment]
Need to check with @rrrlasse.

`is_null` on links behaviour changed in the latest release. before if the link is null, `is_null(something.link.string)` will return false. Now it returns true.

**Comment by bdash** [comment]
@beeender, that behavior was intentionally changed as part of realm/realm-core#1161. It looks like I neglected to mention it in core's release notes, but it is called out in the commit message. Without the behavior change you end up with the unintuitive situation where a query like "link.foo == null && link.foo != null" doesn't include the full result set.

**Comment by beeender** [comment]
@bdash Thanks a lot for the clarification! I am changing the test cases right now!

**Comment by cmelchior** [comment]
retest this please

**Comment by cmelchior** [comment]
Review please @realm/java

File: realm/src/androidTest/java/io/realm/RealmQueryTest.java [hunk 1]
```diff
@@ -1072,35 +1072,36 @@ public void testSumValueForAllNullColumns() {
     public void testIsNullOnLinkField() {
         TestHelper.populateTestRealmForNullTests(testRealm);
 
+        // For the link with null value, query isNull on it's fields should return true.
         // 1 String
-        assertEquals(1, testRealm.where(NullTypes.class).isNull(
+        assertEquals(2, testRealm.where(NullTypes.class).isNull(
```
Comment (kneth) [inline realm/src/androidTest/java/io/realm/RealmQueryTest.java]: Do we have to update the documentation? It seems that `isNull()` is behaving differently now (breaking change?).

File: realm-jni/src/tablequery.hpp [hunk 2]
```diff
@@ -28,7 +28,7 @@ class TableQuery : public realm::Query {
     std::vector<size_t> subtables;  // holds subtable column indices
 
 public:
-    TableQuery(const Query& copy) : realm::Query(copy, realm::Query::TCopyExpressionTag{}) {};
```
Comment (kneth) [inline realm-jni/src/tablequery.hpp]: What is the implication? Is the `realm:Query` object deep-copied or not?

File: realm/src/androidTest/java/io/realm/RealmQueryTest.java (hunk 1 above)
Comment (beeender) [inline realm/src/androidTest/java/io/realm/RealmQueryTest.java]: This detail is not mentioned before... So we can update the document in another PR?

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 3]
```diff
@@ -2038,6 +2039,8 @@ public void onChange() {
                     fail();
                 } finally {
                     bgRealm.close();
+                    bgClosedLatch.countDown();
+
```
Comment (zaki50) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: unnecessary empty line

**Comment by cmelchior** [comment]
Updated with comments @realm/java

File: changelog.txt [hunk 4]
```diff
@@ -1,4 +1,7 @@
-0.83.1
+0.84.0
```
Comment (emanuelez) [inline changelog.txt]: Should't this stay as 0.83.1?

**Comment by kneth** [comment]
:+1:

File: realm/src/androidTest/java/io/realm/RealmQueryTest.java [hunk 5]
```diff
@@ -1072,35 +1072,36 @@ public void testSumValueForAllNullColumns() {
     public void testIsNullOnLinkField() {
         TestHelper.populateTestRealmForNullTests(testRealm);
 
+        // For the link with null value, query isNull on it's fields should return true.
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmQueryTest.java]: it's -> its

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 6]
```diff
@@ -2013,12 +2013,13 @@ public void run() {
         }
     }
 
-    public void testProcessLocalListenersAfterRefresh() {
+    public void testProcessLocalListenersAfterRefresh() throws InterruptedException {
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: I guess we could just catch Exception here

**Comment by zaki50** [comment]
:+1:

File: changelog.txt (hunk 4 above)
Comment (cmelchior) [inline changelog.txt]: Yes, but we need to rewrite the changelog before a release anyway. As this PR must be rebased on top of 0.83.0. For merging to master I think 0.84.0 is the correct version?

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 6 above)
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: You mean throws? Any exception not caught will fail the unit test anyway, so we need only to declare any checked exceptions?

**Comment by emanuelez** [comment]
:+1: apart from `it's -> its`

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 6 above)
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Not important. I should even have mentioned this. :)

File: changelog.txt [hunk 7]
```diff
@@ -1,4 +1,7 @@
...
+ * Updated Realm core to version 0.94.1.
+   - Fixed a bug when using Realm.compactRealm().
```
Comment (bmunkholm) [inline changelog.txt]: Expand description: How could one notice that bug?

**Pull request merged**